### PR TITLE
Add shim to accommodate refactoring.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This is to create a bit of a time buffer for users of Numba internals (as discovered here https://github.com/numba/numba/issues/5214) to make the move to the new Numba code base layout. Anything that was on the summary list here https://github.com/numba/numba/issues/5214#issuecomment-593611360 has been re-exported irrespective of whether it's a good idea or something that will be supported in the future.

Discussion items:
1. What sort of warning should be given to users importing from legacy locations? Some deprecation warning with messaging along the lines of "This module (X) has moved to location (Y), please update your code. Numba 0.50 will not support import from the currently attempted location".
2. Should 1. be switch-off-able via `numba.config`/environment variable? Customers/downstream users and projects of libraries consuming Numba internals probably won't want to see this warning/it has no benefit to them.
3. General discussion about what is reasonable to assume as public/stable. Perhaps for another time/PR.
4. How to test that the shim is ok?! Ask users/use numba-integration-testing?

**Comment by stuartarchibald** [comment]
Marking ready for review, discussion items need resolving as part of the process.

**Comment by eric-wieser** [comment]
Is it worth trying something like using module `__getattr__`s to `DeprecationWarning` about accessing through the shim?

**Comment by esc** [comment]
1.  yes
2. probably
3. yeah, that is a whole 'nother discussion altogether.
4. yes, i'll create a test branch to test against this PR now to help with that.

**Comment by stuartarchibald** [comment]
Note to self. Update deprecation docs.

**Comment by esc** [comment]
Preliminary shim testing is up here: https://github.com/numba/numba-integration-testing/pull/32

**Comment by stuartarchibald** [comment]
Thanks @esc.

**Comment by esc** [comment]
Quick report from the first round of testing: `umap` continues to be unaffected and is good and `pydata-sparse` will probably be O.K. with the shim.

**Comment by esc** [comment]
`fastparquet` and `librosa` will probably be fine also and `pygbm` is back to segfaulting (no more import errors).

**Comment by asmeurer** [comment]
Won't `__getattr__` only work if people import the module and then access names as attributes? It would probably be better to just add the deprecation to the top level of the file so it always happens when you import it.

**Comment by eric-wieser** [comment]
@asmeurer: it's also triggered by a direct import of those names. But yes, perhaps in this case a module-level warning is perfectly adequate.

**Comment by stuartarchibald** [comment]
I think that doing both may be necessary, given removal of this patch in 0.50 will break people's non-updated code when removed, I'm inclined for it to be noisy.

**Comment by esc** [comment]
It looks like awkward is still failing over (https://github.com/numba/numba-integration-testing/pull/34):

```
==================================== ERRORS ====================================
_____________ ERROR collecting tests/test_0118-numba-cpointers.py ______________
tests/test_0118-numba-cpointers.py:15: in <module>
    awkward1_numba_arrayview = pytest.importorskip("awkward1._numba.arrayview")
awkward1/_numba/arrayview.py:77: in <module>
    class LookupModel(numba.datamodel.models.StructModel):
E   AttributeError: module 'numba' has no attribute 'datamodel'
```

**Comment by eric-wieser** [comment]
Seems downstream projects sometimes assume these modules are preimported

**Comment by stuartarchibald** [comment]
I propose adding something like:  https://github.com/stuartarchibald/numba/commit/f5744a2c95fe20c0247f1b2e1fb1601574fa7b08

With that enabled, this:
```python
from numba import analysis
from numba.analysis import ir_extension_usedefs
```
produces:
```
$ python test_shim1.py 
test_shim1.py:1: NumbaDeprecationWarning: An import of was requested from a module that has moved location.
Import requested from: 'numba.analysis', please update to use 'numba.core.analysis' or pin to Numba version 0.48.0.
  from numba import analysis
test_shim1.py:2: NumbaDeprecationWarning: An import of was requested from a module that has moved location.
Import requested from: 'numba.analysis', please update to use 'numba.core.analysis' or pin to Numba version 0.48.0.
  from numba.analysis import ir_extension_usedefs
```

**Comment by eric-wieser** [comment]
That change isn't quite correct, but I think the idea is sound. If you ping me when you turn that into a PR I'll review it.

**Comment by stuartarchibald** [comment]
> That change isn't quite correct, but I think the idea is sound. If you ping me when you turn that into a PR I'll review it.

Thanks for looking, I'm going to pull that onto this PR, so please just comment on the patch and I'll make changes. I really want this all to go in as one patch so it's trivial to revert in 0.50. Thanks!

File: numba/analysis.py [hunk 1]
```diff
@@ -1,2 +1,12 @@
-from numba.core.analysis import * # noqa: F403, F401
-from numba.core.analysis import _use_defs_result # noqa: F401
+from numba.core.errors import (deprecate_moved_module,
+                               deprecate_moved_module_getattr)
```
Comment (eric-wieser) [inline numba/analysis.py]: Perhaps wise not to expose these properties in auto-complete:
```suggestion
import numba.core.errors as _errors
```

File: numba/analysis.py [hunk 2]
```diff
@@ -1,2 +1,12 @@
...
+from numba.core.utils import PYVERSION
+
+if PYVERSION > (3, 6):
+    __getattr__ = deprecate_moved_module_getattr(__name__,
+                                                 'numba.core.analysis')
+else:
+    from numba.core.analysis import * # noqa: F403, F401
+    from numba.core.analysis import _use_defs_result # noqa: F401
+
+deprecate_moved_module(__name__, 'numba.core.analysis')
```
Comment (eric-wieser) [inline numba/analysis.py]:
```suggestion
_errors.deprecate_moved_module(__name__, 'numba.core.analysis')
```

File: numba/analysis.py [hunk 3]
```diff
@@ -1,2 +1,12 @@
-from numba.core.analysis import * # noqa: F403, F401
-from numba.core.analysis import _use_defs_result # noqa: F401
+from numba.core.errors import (deprecate_moved_module,
+                               deprecate_moved_module_getattr)
+from numba.core.utils import PYVERSION
+
+if PYVERSION > (3, 6):
```
Comment (eric-wieser) [inline numba/analysis.py]: Or just:
```suggestion
import sys

if sys.version_info >= (3, 7):
```

File: numba/analysis.py [hunk 4]
```diff
@@ -1,2 +1,12 @@
...
+    __getattr__ = deprecate_moved_module_getattr(__name__,
+                                                 'numba.core.analysis')
+else:
+    from numba.core.analysis import * # noqa: F403, F401
+    from numba.core.analysis import _use_defs_result # noqa: F401
```
Comment (eric-wieser) [inline numba/analysis.py]: PEP8 wants two spaces
```suggestion
    from numba.core.analysis import *  # noqa: F403, F401
    from numba.core.analysis import _use_defs_result  # noqa: F401
```

File: numba/analysis.py (hunk 3 above)
Comment (eric-wieser) [reply numba/analysis.py]: `>=` is probably better here even if you keep using `PYVERSION`

File: numba/typeconv.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from numba.core.typeconv import * # noqa: F403, F401
```
Comment (eric-wieser) [inline numba/typeconv.py]: All the files below this line need deprecations too, right?

Comment (stuartarchibald) [reply numba/typeconv.py]: yes, I pushed up this patch just to see what CI thought :) not proposing for one instant that it is actually used as it's way off being finished.

File: numba/analysis.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/analysis.py]: thanks, fixed

File: numba/analysis.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/analysis.py]: Agree, fixed. Also, for future, Numba uses `PYVERSION`.

Comment (eric-wieser) [reply numba/analysis.py]: Any reason you have `PYVERSION == sys.version_info[:2]` and not just `PYVERSION = sys.version_info`?

Comment (stuartarchibald) [reply numba/analysis.py]: Historical, don't think there's any particular need.

File: numba/core/errors.py [hunk 6]
```diff
@@ -392,6 +393,38 @@ def wrapper(*args, **kwargs):
...
+
+def deprecate_moved_module_getattr(old_module, new_module):
+    """For replacing module level __getattr__ to warn users above modules moving
+    locations
+    """
+    def wrapper(attr):
+        mod = import_module(new_module)
+        ret_attr = getattr(mod, attr)
+        warnings.warn(_moved_msg2.format(attr, old_module, new_module),
+                      category=NumbaDeprecationWarning, stacklevel=2)
```
Comment (eric-wieser) [inline numba/core/errors.py]:
```suggestion
                      category=NumbaDeprecationWarning, stacklevel=3)
```

File: numba/core/errors.py [hunk 7]
```diff
@@ -392,6 +393,45 @@ def wrapper(*args, **kwargs):
         return decorator
 
 
+_moved_msg1 = ("An import was requested from a module that has moved location."
+               "\nImport requested from: '{}', please update to use "
+               "'{}' or pin to Numba version 0.48.0.")
```
Comment (stuartarchibald) [inline numba/core/errors.py]: Add `This alias will be removed in Numba 0.50.0.`

**Comment by stuartarchibald** [comment]
https://github.com/stuartarchibald/numba/tree/wip/refactor_shim_1_backup is the same as this branch, am going to flatten all this into one patch now. 9858302 will get force pushed to a single patch.

File: numba/core/errors.py (hunk 7 above)
Comment (stuartarchibald) [reply numba/core/errors.py]: I went with "This alias will not be present in Numba version 0.50.0.".

**Pull request merged**